core/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -0,0 +1,172 @@
+        listener.accept(true);
+    }
+
+    /**
+     * Setup the translog used to find the last refreshed location.
+     */
+    public void setTranslog(Translog translog) {
+        this.translog = translog;
+    }
+
+    // Implementation of ReferenceManager.RefreshListener that adapts Lucene's RefreshListener into Elasticsearch's refresh listeners.
+    private Translog translog;
+    /**
+     * Snapshot of the translog location before the current refresh if there is a refresh going on or null. Doesn't have to be volatile
+     * because when it is used by the refreshing thread.
+     */
+    private Translog.Location currentRefreshLocation;
+
+    @Override
+    public void beforeRefresh() throws IOException {
+        currentRefreshLocation = translog.getLastWriteLocation();
+    }
+
+    @Override
+    public void afterRefresh(boolean didRefresh) throws IOException {
+        // This intentionally ignores didRefresh so a refresh call over the API can force rechecking the refreshListeners.
+        if (null == currentRefreshLocation) {
+            /*
+             * The translog had an empty last write location at the start of the refresh so we can't alert anyone to anything. This
+             * usually happens during recovery. The next refresh cycle out to pick up this refresh.
+             */
+            return;
+        }
+        /*
+         * First set the lastRefreshedLocation so listeners that come in locations before that will just execute inline without messing
+         * around with refreshListeners at all.
+         */
+        lastRefreshedLocation = currentRefreshLocation;
+        /*
+         * Grab the current refresh listeners and replace them with a new list while synchronized. Any listeners that come in after this
+         * won't be in the list we iterate over and very likely won't be candidates for refresh anyway because we've already moved the
+         * lastRefreshedLocation.
+         */
+        List<Tuple<Translog.Location, Consumer<Boolean>>> candidates;
+        synchronized (this) {
+            candidates = refreshListeners;
+            // No listeners to check so just bail early
+            if (candidates == null) {
+                return;
+            }
+            refreshListeners = null;
+        }
+        /*
+         * Iterate the list of listeners, preserving the ones that we couldn't fire in a new list. We expect to fire most of them so this
+         * copying should be minimial. Much less overhead than removing all of the fired ones from the list.
+         */
+        List<Tuple<Translog.Location, Consumer<Boolean>>> preservedListeners = null;
+        for (Tuple<Translog.Location, Consumer<Boolean>> tuple : candidates) {
+            Translog.Location location = tuple.v1();
+            Consumer<Boolean> listener = tuple.v2();
+            if (location.compareTo(currentRefreshLocation) <= 0) {
+                listenerExecutor.execute(() -> listener.accept(false));
```

Comment:
this looks all great - I just wonder if it would make sense to batch the  accept calls up and only one one executor to fire them all up? I think that would make the entire thing more efficient and would use less resources on the threadpool since we only spawn one?